[week1] Java를 이용한 ToDo REST API 만들기 - kwakcena

Kwakcena opened the pull request:
과제를 제출합니다.

자바 언어에 익숙하지 않아서 코드가 많이 엉망인 것 같습니다 ㅠ 
일단 동작하게 하고 점점 리팩토링을 통해서 개선 해 나가겠습니다.
![스크린샷 2021-12-03 오전 2 42 07](https://user-images.githubusercontent.com/45390172/144475025-bbb6963f-3783-4afd-a570-68078c9bc43a.png)

## TODO
- [X] `GET /tasks` 요청 시 전체 tasks 응답하기
- [X] `GET /tasks/{id}` 요청 시 id와 일치하는 task 응답하기
- [X] `POST /tasks` 요청 시 요청에 포함한 task를 추가하기
- [X] `PUT /tasks/{id}` 요청 시 id와 일치하는 task를 update 하기
- [X] `DELETE /tasks/{id}` 요청 시 id와 일치하는 task를 찾아서 삭제하기
- [ ] 요청이 실패했을 때 처리하기
- [ ] 중복되는 코드 제거하기

## 어떤 고민을 했는가
- `DELETE` 요청 시 응답으로 반환하는 content는 보통 삭제한 내용인가? 아니면 삭제 성공했다 true, false?
- `PUT` 요청을 통해 기존의 데이터를 수정해야 할 때 객체 내부의 데이터를 어떻게 수정하는게 좋을까
- 현재는 `uri path`에서 `id`를 읽는 방법이 정규식을 이용하여 있는지 없는지에 따라 마지막 `/` 를 찾아 반환하고 있는데, 좀 더 깔끔한 방법은 없을까
- `handler` 클래스 안에서 많은 일들이 일어나고 있는데, 어떻게 쪼개야 잘 했다고 소문이 날까. 무엇을 기준으로 해야 할까

Kwakcena committed:
http server를 만들고 실행하라

8001번 포트를 사용하는 간단한 httpServer를 만들었습니다.
현재는 context에 handler가 등록되어있지 않아서 500 server error를
응답합니다.

Kwakcena committed:
Todo http handler를 만들어라

Todo REST API 의 handler를 만들고 등록합니다.

Kwakcena committed:
handler에서 method, path를 출력하라

TodoHttpHandler에서 요청받은 method와 path를 출력합니다.

Kwakcena committed:
Response header를 설정하라

요청의 응답에 header를 지정합니다.

현재는 200 을 보내고 content의 length와 같이 보내고 있습니다.

Kwakcena committed:
GET 요청 시 빈 배열을 반환한다

GET 요청 시 content에 빈 배열의 문자열을 반환합니다.
Output stream을 지정하고 content를 써서 반환하도록 하였습니다.

Kwakcena committed:
InputStream을 읽어서 출력하라

요청으로 받은 Body 값을 읽어서 터미널에 출력합니다.

Kwakcena committed:
handler의 위치를 수정합니다

handler가 models에 있는게 맞지 않아서 이를 수정합니다.

Kwakcena committed:
Task modle을 만들어라

Task class를 만들었습니다.

Kwakcena committed:
jackson을 이용하여 task를 반환하라

jackson을 이용하여 tasks의 데이터를 json 형태로 변환하여 outputstream을
형성하고 이를 반환합니다.

Kwakcena committed:
GET /tasks 요청 시 빈 배열 반환하라

처음에 tasks가 비어있으면 빈 배열을 반환합니다.

Kwakcena committed:
POST /task 를 통해 task를 추가할 수 있다

POST 요청을 통해 task를 추가할 수 있습니다.

Kwakcena committed:
반환 content는 모두 TaskToJson을 사용한다

GET, POST 모두 반환하는 문자열은 Task를 JSON으로 변환한 데이터이기
때문에 분기를 제거하고 content에 TaskToJSON() 만 사용하도록 합니다.

Kwakcena committed:
Task id로 상세 조회한 task를 응답하라

GET /tasks/{id} 요청에 따라 id가 있을 경우 tasks에서 해당 id의 task를 찾아 반환하는 로직을 추가하였습니다.
또한 모든 tasks를 반환하는 함수를 TasksToJSON() 으로 수정하고, 하나의 tasks를 json으로 변환하는 함수를
TaskToJSON()으로 하였습니다.

요청의 path에 id 존재 여부는 정규식을 이용하였고, 정규식이 일치할 경우 path에서 id를 Long 타입으로 변환하여
이를 반환합니다.

Kwakcena committed:
POST 요청에 대한 처리를 추가하라

task 추가에 대한 POST 요청이 들어오면 task를 추가하고 방금 추가한 task를 content로 반환합니다.

Kwakcena committed:
DELETE 요청 시 task 를 삭제한다

DELETE 요청 시 task를 삭제합니다. 요청 시 보내는 task id와 일치하는 task를
tasks에서 찾아 ArrayList의 remove 메소드를 이용하여 제거합니다.

Kwakcena committed:
PUT 요청이 오면 task 를 수정할 수 있다

PUT 요청이 왔을 때 task를 수정합니다.
자바스크립트처럼 기존 객체에서 변경할 프로퍼티만 변경시킬 수 있는 방법이 떠오르지 않아
변경 전 task의 id를 update 할 task에 넣어줬습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,10 +1,27 @@
 
     public static void main(String[] args) {
+        try {
+            InetSocketAddress address = new InetSocketAddress(8001);
```
8001이 의미하는 것이 무엇일까요? Magic literal로 분리하는 것은 어떨까요

hannut91 commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,107 @@
+        String body = new BufferedReader(new InputStreamReader(inputStream))
+                .lines().collect(Collectors.joining("\n"));
+
+        String content = TasksToJSON();
```
Method의 이름은 camelCase로 작성하는게 좋을 것 같아요

hannut91 commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,107 @@
+                return task;
+            }
+        }
+        return null;
```
어떤 함수가 null을 반환하는 것은 권장하지 않습니다. 왜냐하면 이 함수를 사용하는 입장에서 null올 것이라는 것을 기대하기 어렵기 때문이에요. 함수의 signature를 보아도 추측할 수 없고, 이 함수를 사용하는 사람에게 이 함수는 null을 반환하니 꼭 null을 처리해야해! 라고 주석으로 작성하거나 함수를 보기 전까지는 알기 어렵기 때문입니다

hannut91 commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,107 @@
+            content = TaskToJSON(task);
+        } else if(method.equals("POST") && !body.isBlank()) {
+            Task task = toTask(body);
+            task.setId(id++);
```
메서드의 값을 전달할 때 연산의 결과를 전달하는게 더 좋을 것 같아요

hannut91 commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,107 @@
+
+        if(method.equals("GET") && taskId != null) {
+            Task task = findTaskById(taskId);
+            content = TaskToJSON(task);
```
위에서 `TaskToJSON`를 호출하고 있는데 또 호출하고 있네요

Kwakcena committed:
[풀이] send 메서드를 추출하라

응답의 header를 설정하고 outputstream 을 만들어 응답을 보내는 부분을
send 메서드로 추출합니다.

Kwakcena committed:
[풀이] body를 가져오는 getBody 메서드를 만들어라

요청으로 들어온 body를 읽는 메소드인 getBody를 만들었습니다.

Kwakcena committed:
[풀이] tasksToJSON과 taskToJSON를 하나로 합친다

task 하나를 반환하는 메소드는 taskToJSON으로 만들었고, 전체 할일 목록을
반환하는 메소드는 tasksToJSON으로 만들었는데,

파라미터로 Object 를 주면 이를 하나로 합칠 수 있었다.

Kwakcena committed:
[풀이] content 변수를 걷어내고 send를 사용한다

content 변수를 걷어내고 각각의 요청 method에 send를 적절하게 사용합니다.

Kwakcena committed:
[풀이] handleCollection 을 구현합니다

GET과 POST 요청에 대한 처리를 수행하는 handleCollection 메소드를
만들었습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,107 @@
+        return outputStream.toString();
+    }
+
+    private Long parseTaskId(URI uri) {
+        String path = uri.getPath();
+        String pattern = "^\\/tasks\\/\\d+$";
+
+        if(!Pattern.matches(pattern, path)) {
+            return null;
+        }
+
+        try {
+            Long id = Long.parseLong(path.substring(path.lastIndexOf('/') + 1));
+            return id;
+        } catch (Exception e) {
+            return null;
+        }
+    }
```
정규식을 썼는데, 다시  substring으로 값을 추출하는게 이상하네요. 정규식을 사용해서 id를 추출해 보세요

Kwakcena replied:
[91ba580](https://github.com/CodeSoom/spring-week1-assignment-1/pull/68/commits/91ba580b057f5c7e0cc63759fc2e64adef43d612) 에서 정규식으로 id를 추출 해 봤습니다.

Kwakcena committed:
[풀이] handleItem 메소드를 구현합니다

GET, DELETE, PUT 같은 요청은 하나의 아이템에 대한 요청을 수행할 수
있습니다.

따라서 이를 처리하는 handleItem 메소드를 만듭니다.

Kwakcena committed:
[풀이] findTask를 refactoring 합니다.

findTask를 refactoring 합니다.

Kwakcena committed:
[풀이] 사용하지 않는 메소드 제거

사용하지 않는 메소드를 제거합니다.

Kwakcena committed:
[풀이] task가 없는 경우 404 에러를 반환한다

id가 잘못 입력되면 findTask에서 null을 반환하기 때문에 404 에러를
반환하는 처리를 추가합니다.

Kwakcena committed:
[풀이] task 추가 시 id를 생성하는 로직 추가

generateId 메소드를 만들어서 id를 추가하는 로직을 만듭니다.

Kwakcena committed:
[풀이] PATCH를 추가합니다

update에는 put 외에 patch도 있으므로 이를 추가하고,
post 요청의 성공 응답은 201 코드이므로 이를 수정합니다.

Kwakcena committed:
[풀이] update 요청의 로직을 수정하라

body로 들어온 데이터는 source 이므로 이름을 수정했고,
이의 title을 현재 task의 title로 수정하여 응답으로 보낸다.

Kwakcena committed:
[풀이] handler를 구현합니다

각각의 요청에 맞는 handler로 메소드를 만듭니다.

Kwakcena committed:
handleCollection에서 잘못된 요청이 올 경우 처리

handleCollection은 GET 이나 POST 에 대한 요청을 수행하는데,
만약 요청이 다음과 같이 온다면
http PUT localhost:8001/tasks

아무런 동작을 하지 않고 기다린다.

이런 경우를 대비해서 404를 반환하게 하였다.

Kwakcena committed:
Magic literal 로 PORT를 분리합니다

PORT를 magic literal로 뺐습니다.

Kwakcena committed:
정규식을 이용하여 ID를 추출하라

정규식을 이용하여 요청에 포함되어 있는 id를 추출합니다.

Kwakcena commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -29,12 +31,22 @@ public void handle(HttpExchange exchange) throws IOException {
     }
 
+    private Long getTaskId(String path) {
+        Pattern pattern = Pattern.compile("^\\/tasks\\/(\\d+)$");
+        Matcher matcher = pattern.matcher(path);
+
+        if(matcher.find()) {
+            return Long.parseLong(matcher.group(1));
```
정규식 `"^\\/tasks\\/(\\d+)$"` 에 대해서 matcher를 사용하면 소괄호로 묶은 패턴에 일치하는 문자열 값을 구할 수 있습니다.
만약 `/tasks/123` 과 같이 요청이 온다면 다음과 같은 문자열 결과를 반환합니다.

`matcher.group()` 또는 `matcher.group(0)`  => `/tasks/123`
`matcher.group(1)` => `123`

hannut91 commented:
> 핵심은 문제를 작게 나누고 작은 문제 하나씩에 집중하면서 해결해 나가는 과정을 보여주신 것 같다.

잘 캐치해내신 것 같네요. 잘 관찰하여 많이 배우셨으면 좋겠습니다. 회괴를 작성할 때 앞으로 어떤 것을 할 것인지도 적어보시면 좋을 것 같아요.
단순하게 다음 과제에서는 무엇을 중점적으로 학습할지 결정하는 것도 좋을 것 같습니다. 고생 많으셨습니다
